core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,322 @@ class DAGSchedulerSuite
+
+    assert(scheduler.runningStages.nonEmpty)
+    assert(!ended)
+
+    // Confirm job finished succesfully
+    val stage1Attempt2 = taskSets.last
+    val completions_1_2 = stage1Attempt2.tasks.zipWithIndex.map(_ => (Success, 42)).toSeq
+
+    complete(stage1Attempt2, completions_1_2)
+    sc.listenerBus.waitUntilEmpty(1000)
+    assert(ended === true)
+    assert(results === Map(1 -> 42))
+  }
+
+  /**
+   * In this test we demonstrate that only consecutive failures trigger a stage abort. A stage may
+   * fail multiples, succeed, then fail a few more times (because its run again by downstream
+   * dependencies). The total number of failed attempts for one stage will go over the limit,
+   * but that doesn't matter, since they have successes in the middle.
+   */
+  test("Abort should only trigger after consecutive stage failures") {
+    // Create a new Listener to confirm that the listenerBus sees the JobEnd message
+    // when we abort the stage. This message will also be consumed by the EventLoggingListener
+    // so this will propagate up to the user.
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
+    val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, null)
+    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne)).cache()
+    val shuffleDepTwo = new ShuffleDependency(shuffleTwoRdd, null)
+    val finalRdd = new MyRDD(sc, 1, List(shuffleDepTwo))
+    submit(finalRdd, Array(0))
+
+    for (attempt <- 0 until Stage.MAX_STAGE_FAILURES-1) {
+      println(s"attempt = $attempt")
+      println(taskSets.mkString(","))
+
+      // Make each task in stage 0 success
+      val stage0Attempt = taskSets.last
+
+      val completions = makeCompletions(stage0Attempt);
+
+      // Run stage 0
+      complete(stage0Attempt, completions)
+
+      // Now we should have a new taskSet, for a new attempt of stage 1.
+      // We will have one fetch failure for this task set
+      val stage1Attempt = taskSets.last
+      val stage1Successes =
+        stage1Attempt.tasks.tail.map { _ => (Success, makeMapStatus("hostB", 1))}
+
+      // Run Stage 1, this time with a task failure
+      complete(stage1Attempt,
+        Seq(
+          (FetchFailed(makeBlockManagerId("hostA"), shuffleDepOne.shuffleId, 0, 0, "ignored"), null)
+        ) ++ stage1Successes
+      )
+
+      scheduler.resubmitFailedStages()
+
+      // Confirm we have not yet aborted
+      assert(scheduler.runningStages.nonEmpty)
+      assert(!ended)
+    }
+
+    // Now succeed stage 1 and fail stage 2
+    val stage1attempt4 = taskSets.last
+    val completions_1_4 = makeCompletions(stage1attempt4)
+
+    // Succeed 1
+    complete(stage1attempt4, completions_1_4)
```

Comment:
complete the retry for stage 0 first.  Also, lets stick to 0-indexed naming for the stage attempts (sorry I know I use 1-index in the English discussion, not sure of a good way around the confusion ...)

```scala
// Now succeed stage 1 and fail stage 2
val stage0Attempt3 = taskSets.last
checkStageId(0, 3, stage0Attempt3)
complete(stage0Attempt3, makeCompletions(stage0Attempt3, 2))

val stage1attempt3 = taskSets.last
checkStageId(1, 3, stage1attempt3)
complete(stage1attempt3, makeCompletions(stage1attempt3, 1))
```